dubbo-cluster/src/test/java/com/alibaba/dubbo/rpc/cluster/loadbalance/LoadBalanceTest.java

```diff
@@ -120,80 +245,120 @@ public void testRoundRobinLoadBalance_select() {
-        System.out.println(sumInvoker1);
-        System.out.println(sumInvoker2);
-        Assert.assertEquals("select failed!", sumInvoker1 + sumInvoker2, loop);
+        int runs = 10000;
+        //temporarily add a new invoker
+        weightInvokers.add(weightInvokerTmp);
+        try {
+            Map<Invoker, InvokeResult> resultMap = getWeightedInvokeResult(runs, RoundRobinLoadBalance.NAME);
+            assertStrictWRRResult(runs, resultMap);
+            RoundRobinLoadBalance lb = (RoundRobinLoadBalance)getLoadBalance(RoundRobinLoadBalance.NAME);
+            Assert.assertEquals(weightInvokers.size(), lb.getInvokerAddrList(weightInvokers, weightTestInvocation).size());
+            
+            //remove the tmp invoker and retry, should recycle its cache
+            weightInvokers.remove(weightInvokerTmp);
+            resultMap = getWeightedInvokeResult(runs, RoundRobinLoadBalance.NAME);
+            assertStrictWRRResult(runs, resultMap);
+            Assert.assertEquals(weightInvokers.size(), lb.getInvokerAddrList(weightInvokers, weightTestInvocation).size());
+        } finally {
+            weightInvokers.remove(weightInvokerTmp);
+        }
     }
 
     @Test
-    public void testSelectByWeightRandom() {
-        int sumInvoker1 = 0;
-        int sumInvoker2 = 0;
-        int sumInvoker3 = 0;
+    public void testSelectByWeightLeastActive() {
         int loop = 10000;
-        RandomLoadBalance lb = new RandomLoadBalance();
-        for (int i = 0; i < loop; i++) {
-            Invoker selected = lb.select(weightInvokers, null, weightTestInvocation);
-            if (selected.getUrl().getProtocol().equals("test1")) {
-                sumInvoker1++;
+        Map<Invoker, InvokeResult> resultMap = getWeightedInvokeResult(loop, LeastActiveLoadBalance.NAME);
+        int test3Weight = 0;
+        //find test3's weight
+        for (Entry<Invoker, InvokeResult> entry : resultMap.entrySet()) {
+            String protocol = entry.getKey().getUrl().getProtocol();
+            if (protocol.equals("test3")) {
+                //should no request invoked to test3
+                Assert.assertEquals(0, entry.getValue().getCount().get());
+                test3Weight = entry.getValue().getWeight();
             }
-            if (selected.getUrl().getProtocol().equals("test2")) {
-                sumInvoker2++;
-            }
-            if (selected.getUrl().getProtocol().equals("test3")) {
-                sumInvoker3++;
+        }
+        for (Entry<Invoker, InvokeResult> entry : resultMap.entrySet()) {
+            String protocol = entry.getKey().getUrl().getProtocol();
+            if (!protocol.equals("test3")) {
+                InvokeResult invokeResult = entry.getValue();
+                invokeResult.setTotalWeight(invokeResult.getTotalWeight() - test3Weight);
+                // less then 12% requests to expected
+                Assert.assertTrue(invokeResult.getDeltaPercentage(loop) < 12);
```

Comment:
I don't think this is a good way to calc percentage of select.
It is not stable.

Reply:
> I don't think this is a good way to calc percentage of select.
> It is not stable.

The Unit Test before only covers the invokers without weight specified. So if we add weight to them the actually result will be indetermination in `Random` etc. . So i tried to use percentage to make it.

So is there other way to do that? The old logics just print the result.

Reply:
There is no better way yet. But using percentages is not a good way, because it is not stable, which causes us to spend a lot of time re-executing the entire ci.

Reply:
> There is no better way yet. But using percentages is not a good way, because it is not stable, which causes us to spend a lot of time re-executing the entire ci.

I gave a big threshold for those tests and actually there must be a problem if they fail under this condition.

Is it better then no assertion at all?

And Maybe we will optimize them in future to get a better threshold(eg. better Random).

Reply:
If the threshold is too large, we still can't find it if there is a problem. So how do we determine a suitable threshold?
I think this is difficult. And even if we find this threshold, ut can't be stable, so I don't think we should handle this.

How do u think about it?

Reply:
> If the threshold is too large, we still can't find it if there is a problem. So how do we determine a suitable threshold?
> I think this is difficult. And even if we find this threshold, ut can't be stable, so I don't think we should handle this.
> 
> How do u think about it?

Currently the threshold is 12% diff to expected requests should dispatch. Maybe there's some problem in Random for example, it will between 0~8% (I think the result should be stable when we simulate 10k request).

12% under weight is acceptable to me(more like a complete unit test and reliable after hundreds loop), but your opinion is reasonable too.

So i won't consist on it and we can make a decision whether to leave them as they were.

btw the CI failed and i am checking the problem(seems not because this unit test).